Electron/RON/components/PingTool.tsx

```diff
@@ -0,0 +1,248 @@
+import React, { Component } from "react"
+import CSS from "csstype"
+import { exec } from "child_process"
+import { rovecomm } from "../../Core/RoveProtocol/Rovecomm"
+import { DATAID, NetworkDevices } from "../../Core/RoveProtocol/RovecommManifest"
+import { ColorStyleConverter } from "../../Core/ColorConverter"
+
+const h1Style: CSS.Properties = {
+  fontFamily: "arial",
+  fontSize: "18px",
+  fontWeight: "bold",
+  textAlign: "center",
+  margin: "5px 0px",
+}
+const container: CSS.Properties = {
+  display: "flex",
+  flexDirection: "column",
+  fontFamily: "arial",
+  width: "640px",
+  borderTopWidth: "28px",
+  borderColor: "#990000",
+  borderBottomWidth: "2px",
+  borderStyle: "solid",
+  padding: "5px",
+  alignItems: "center",
+}
+const label: CSS.Properties = {
+  marginTop: "-10px",
+  position: "relative",
+  top: "24px",
+  left: "3px",
+  fontFamily: "arial",
+  fontSize: "16px",
+  zIndex: 1,
+  color: "white",
+}
+const selectbox: CSS.Properties = {
+  display: "flex",
+  flexDirection: "row",
+  width: "450px",
+  margin: "2.5px",
+  justifyContent: "space-between",
+}
+const labelbox: CSS.Properties = {
+  display: "flex",
+  flexDirection: "row",
+  width: "450px",
+  margin: "2.5px",
+  justifyContent: "space-between",
+  textAlign: "center",
+}
+const auto: CSS.Properties = {
+  width: "10%",
+}
+const name: CSS.Properties = {
+  width: "50%",
+}
+const num: CSS.Properties = {
+  width: "10%",
+  textAlign: "center",
+}
+const but: CSS.Properties = {
+  width: "10%",
+}
+const hid: CSS.Properties = {
+  width: "10%",
+  visibility: "hidden",
+}
+
+interface IProps {}
+
+interface IState {
+  devices: any
+}
+
+// For colorConverter
+const min = 0
+const cutoff = 45
+const max = 100
+const greenHue = 120
+const redHue = 360
+
+class PingTool extends Component<IProps, IState> {
+  constructor(props: any) {
+    super(props)
+    const devices = {}
+    Object.keys(NetworkDevices).forEach(device => {
+      devices[device] = { autoPing: false, ping: -1 }
+    })
+    Object.keys(DATAID).forEach(board => {
+      devices[board] = { autoPing: false, ping: -1 }
+    })
+    this.state = {
+      devices,
+    }
+    this.ICMP = this.ICMP.bind(this)
+    this.Rove = this.Rove.bind(this)
+    this.AutoPing = this.AutoPing.bind(this)
+    this.AutoPingAll = this.AutoPingAll.bind(this)
+    this.StartAutoPing = this.StartAutoPing.bind(this)
+    this.StartAutoPing(1000)
+  }
+
+  ICMP(device: string): void {
+    // If device is not a network device, it must be a board
+    let deviceInfo = NetworkDevices[device]
+    if (deviceInfo === undefined) {
+      deviceInfo = DATAID[device]
+    }
+    const ip = deviceInfo.Ip
+
+    // Ping command is slightly different for windows/unix
+    let pingCommand = ""
+    if (process.platform === "win32") {
+      pingCommand = `ping -n 1 ${ip}`
+    } else {
+      pingCommand = `ping -c 1 ${ip}`
+    }
```

Comment:
I see you use the same ping code in PingMap, figure this will be in a "later retouch" but just noting you can probably break the ping functionality into it's own file if you are using it across components.

Reply:
This will likely be a "later retouch" problem, as the solution isn't easy. Since the ICMP function currently calls this.setState, it cannot be shared as is. We would have to split it into 2 functions, one external function that returns the delay, and then a class function that properly updates the state. Since exec is a childprocess, the value of delay isn't easily returned.